diff --git a/src/modules/users/users.service.ts b/src/modules/users/users.service.ts
index 2df3ec70..818c8599 100644
--- a/src/modules/users/users.service.ts
+++ b/src/modules/users/users.service.ts
@@ -19,7 +19,9 @@ import { GetAllUsersCommand } from '@libs/contracts/commands';
 import { UserEvent } from '@integration-modules/notifications/interfaces';
 
 import { GetUserSubscriptionRequestHistoryQuery } from '@modules/user-subscription-request-history/queries/get-user-subscription-request-history';
+import { RemoveUsersFromNodeEvent } from '@modules/nodes/events/remove-users-from-node';
 import { RemoveUserFromNodeEvent } from '@modules/nodes/events/remove-user-from-node';
+import { AddUsersToNodeEvent } from '@modules/nodes/events/add-users-to-node';
 import { AddUserToNodeEvent } from '@modules/nodes/events/add-user-to-node';
 
 import { NodesQueuesService } from '@queue/_nodes';
@@ -39,6 +41,7 @@ import {
     BulkDeleteUsersByStatusRequestDto,
     BulkUpdateUsersRequestDto,
     BulkAllUpdateUsersRequestDto,
+    RevokeUserSubscriptionBodyDto,
 } from './dtos';
 import { IGetUserByUnique, IGetUsersByTelegramIdOrEmail, IUpdateUserDto } from './interfaces';
 import { GetCachedShortUuidRangeQuery } from './queries/get-cached-short-uuid-range';
@@ -309,19 +312,27 @@ export class UsersService {
 
     public async revokeUserSubscription(
         userUuid: string,
-        shortUuid?: string,
+        dto?: RevokeUserSubscriptionBodyDto,
     ): Promise<TResult<UserEntity>> {
         try {
             const user = await this.userRepository.getPartialUserByUniqueFields(
                 { uuid: userUuid },
-                ['uuid', 'vlessUuid'],
+                ['uuid', 'vlessUuid', 'shortUuid'],
             );
 
             if (!user) return fail(ERRORS.USER_NOT_FOUND);
 
+            let shortUuid = user.shortUuid;
+
+            if (!dto) {
+                shortUuid = this.createNanoId();
+            } else if (!dto.revokeOnlyPasswords) {
+                shortUuid = dto.shortUuid ?? this.createNanoId();
+            }
+
             const updateResult = await this.userRepository.revokeUserSubscription({
                 uuid: user.uuid,
-                shortUuid: shortUuid ?? this.createNanoId(),
+                shortUuid,
                 trojanPassword: this.createPassword(),
                 vlessUuid: this.createUuid(),
                 ssPassword: this.createPassword(),
@@ -537,11 +548,11 @@ export class UsersService {
                 return ok(new BulkOperationResponseModel(0));
             }
 
+            const usersIdsAndHashes = await this.userRepository.getIdsAndHashesByUserUuids(uuids);
+
             const result = await this.userRepository.deleteManyByUuid(uuids);
 
-            await this.nodesQueuesService.startAllNodesWithoutDeduplication({
-                emitter: 'bulkDeleteUsersByUuid',
-            });
+            await this.eventBus.publish(new RemoveUsersFromNodeEvent(usersIdsAndHashes));
 
             return ok(new BulkOperationResponseModel(result));
         } catch (error) {
@@ -607,31 +618,15 @@ export class UsersService {
         internalSquadsUuids: string[],
     ): Promise<TResult<BulkOperationResponseModel>> {
         try {
-            const usersLength = usersUuids.length;
-            const batchLength = 3_000;
+            const userIds = await this.userRepository.getUserIdsByUuids(usersUuids);
 
-            for (let i = 0; i < usersLength; i += batchLength) {
-                const batchUsersUuids = usersUuids.slice(i, i + batchLength);
-                const batchUsersIds = await this.userRepository.getUserIdsByUuids(batchUsersUuids);
+            await this.userRepository.removeUsersFromInternalSquads(userIds);
 
-                if (batchUsersIds.length === 0) {
-                    continue;
-                }
+            await this.userRepository.addUsersToInternalSquads(userIds, internalSquadsUuids);
 
-                await this.userRepository.removeUsersFromInternalSquads(batchUsersIds);
-
-                await this.userRepository.addUsersToInternalSquads(
-                    batchUsersIds,
-                    internalSquadsUuids,
-                );
-            }
-
-            // TODO: finish later
-            await this.nodesQueuesService.startAllNodes({
-                emitter: 'bulkUpdateUsersInternalSquads',
-            });
+            await this.eventBus.publish(new AddUsersToNodeEvent(userIds));
 
-            return ok(new BulkOperationResponseModel(usersLength));
+            return ok(new BulkOperationResponseModel(userIds.length));
         } catch (error) {
             this.logger.error(error);
             return fail(ERRORS.BULK_ADD_INBOUNDS_TO_USERS_ERROR);
